src/pure/List.mo

```diff
@@ -585,8 +586,13 @@ module {
   ///
   /// *Runtime and space assumes that `f` runs in O(1) time and space.
   public func tabulate<T>(n : Nat, f : Nat -> T) : List<T> {
-    func go(at : Nat, n : Nat) : List<T> = if (n == 0) null else ?(f at, go(at + 1, n - 1));
-    go(0, n)
+    var i = 0;
+    var l : List<T> = null;
+    while (i < n) {
+      l := ?(f i, l);
+      i += 1
+    };
+    reverse l
```

Comment:
This is a shame. A clever optimisation would spot the linear heap allocation discipline arising from the recursive algorithm (did somebody say abstract interpretation?) and pre-allocate the list nodes. Then fill them in like an array...

Reply:
Yes, that's called tail-modulo-cons optimization but we don't have it.